addons/account/models/account_move.py

```diff
@@ -3141,84 +3141,110 @@ def _extend_with_attachments(self, attachments, new=False):
-            if decoder:
-                try:
-                    with self.env.cr.savepoint():
-                        invoice = current_invoice or self.create({})
-                        success = decoder(invoice, file_data, new)
+        def process_import(current_invoice, file_data, attachment, passed_file_data_list, invoices, new, no_invoice):
+            if current_invoice:
+                decoder_target = current_invoice
+            else:
+                defaults = no_invoice.default_get(['move_type', 'journal_id'])
+                decoder_target = no_invoice.new(defaults)
 
-                        if success or file_data['attachment'].mimetype in ALLOWED_MIMETYPES:
-                            invoice._link_bill_origin_to_purchase_orders(timeout=4)
-                            invoices |= invoice
-                            current_invoice = self.env['account.move']
-                            add_file_data_results(file_data, invoice)
+            if not (decoder := decoder_target._get_edi_decoder(file_data, new=new)):
+                to_process = []
+            elif parts_details := file_data.pop('parts', []):
+                to_process = [{**file_data.copy(), **part} for part in parts_details]
+            else:
+                to_process = [file_data]
 
+            for content_to_import in to_process:
+                invoice = current_invoice or import_self.create({})
+                try:
+                    if success := (
+                        decoder(invoice, content_to_import, new)
+                        or attachment.mimetype in ALLOWED_MIMETYPES
+                    ):
+                        invoice._link_bill_origin_to_purchase_orders(timeout=4)
+                        invoices |= invoice
+                        current_invoice -= current_invoice
+                        passed_file_data_list.append(content_to_import)
                 except RedirectWarning:
                     raise
                 except Exception:
-                    message = _("Error importing attachment '%s' as invoice (decoder=%s)", file_data['filename'], decoder.__name__)
+                    success = False
+                    message = _(
+                        "Error importing attachment '%(file_name)s' as invoice (decoder=%(decoder)s)",
+                        file_name=file_data['filename'],
+                        decoder=decoder.__name__,
+                    )
                     current_invoice.sudo().message_post(body=message)
                     _logger.exception(message)
-
-            passed_file_data_list.append(file_data)
-            close_file(file_data)
+                finally:
+                    if success:
+                        import_cr.commit()
+                    else:
+                        import_cr.rollback()
+
+        with closing(self.env.registry.cursor()) as import_cr:
```

Comment:
It will only do so if/when it can. Other transactions using savepoints can also prevent cleaning up the memory
More details here
https://github.com/odoo/odoo/pull/71395#issuecomment-851440042

Reply:
This PR is in a messy state.
If I have to doubleloop on parts of a file, then savepoint has to go, and it's not simple. To be true, I'd like to refactor the whole extend with attachments in master, since there's lot of code here just to filter out cases peculiar to one of the three callers (mail, post on invoice, import from journal) lots of continue etc. I asked Laurent some time but it look he s given this up to you Antoine.
The reason to refactor it is precisely to also have control on invoice creation.
In some cases we already have a preliminary function (i.e. create with attachments) so I don't really know why there is, for example, a "new" parameter here to spaghetti it up things a bit, not counting 5 continue s.
This is the core of invoice import, and the level of clarity and readabiliy is quite low.

Reply:
Ideally, I'd like decoders to be generators that only extract dicts of values from the files, instead of manipulating invoices while reading. The creation or update can then just be a load_records, just like we do for chart templates.

The whole idea of creating a fixed number of invoices before reading the files is wrong imho, if you have errors (amounting to 0) or multiple invoices per file makes it difficult.

One of the things that make this passage  hard is that PDFs can contain XMLs that contain PDFs... and we might have to create the invoice from the XML, since it's the most structured data inside the file. If we do, then the rest of attachment (embedding included) must only be attached to the invoice, but no further be examined.

Another, is that if you have multiple attachments from the same mail, only the first that yields an invoice is considered, the rest is just attached.
But this doesn't really be a concern when you import from journal or attach to a message in the chatter

Reply:
I totally agree this would be a great thing to refactor in master (I mentioned it to Laurent and he agreed we need to refactor it - which is maybe why he pinged me? I'm not sure whether he wanted me to just have a look or whether to do a full review though.)

I also realize that removing the savepoint isn't simple. Comparing the various options, none seems perfectly satisfactory:
- savepoints: perf problem
- in the same transaction, commit / rollback after each import: it's not great because it can mess up the request retry mechanism in case of a serialization error, depending on what code was executed before calling the import code. In addition, it means we need some way of keeping track of what was already successfully imported. Or we need to manually catch serialization errors during import and do a rollback + retry in that case, which is messy.
- open new transactions like you suggested: similar problems, as we probably need to do a commit before opening the new transactions, and serialization errors can still occur so we need to handle them.
- delegate each import to a cron job: would solve the concurrency problems, but we don't have enough control over when the crons are executed. Plus, crons aren't designed to be called multiple times in different transactions (they are supposed to batch things as much as possible in a single transaction).

So, my last thought was, we can maybe create our own 'synchronous cron' just for this purpose? An object that allows you to schedule multiple jobs, each to be run in its own transaction, with retry mechanisms in case of serialization errors, and which would be run right when the request returns?

Maybe it would be a lot of development, but to me it seems like that would fit our needs.

Reply:
> Ideally, I'd like decoders to be generators that only extract dicts of values from the files, instead of manipulating invoices while reading. The creation or update can then just be a load_records, just like we do for chart templates.

That would be really nice, if it's feasible. I wonder though whether in practice we need to use the computes to progressively 'build up' an invoice. If we don't, that would definitely be the way to go, I think.

> The whole idea of creating a fixed number of invoices before reading the files is wrong imho, if you have errors (amounting to 0 invoices) or multiple invoices per file it makes things hard.
> 
> One of the things that make this passage hard is that PDFs can contain XMLs that contain PDFs... and we might have to create the invoice from the XML, since it's the most structured data inside the file. If we do, then the rest of attachment (embedding included) must only be attached to the invoice, but no further be examined.
> 
> Another, is that if you have multiple attachments from the same mail, only the first that yields an invoice is considered, the rest is just attached. But this doesn't really be a concern when you import from journal or attach to a message in the chatter

It's funny, I was thinking so too! I think a solution would be to more clearly separate the logic that determines the attachments to decode (including separating embedded XMLs) from the logic that calls the decoders on the right attachments.